### PR TITLE
Add a key management simulator for data protection

### PR DESCRIPTION
**Pull request by amcasey**
It's not properly a sample, since it requires IVT, but it's much easier to read and use as a standalone app than as a test.  (It would also make a poor test, since perfect performance is not expected and thus there's no way to define success.)

Simulates running an app with a given number of instances for a given number of days and computes a score (lower is better) indicating how likely missing key errors were during that simulated time period.  Also tracks calls to the storage and encryption backends so we can see whether e.g. retries have a significant perf impact.

In the future, it would be nice if instances started and stopped over the course of the run, rather than all living for the entire duration.

### CODE REVIEW

File: src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs [hunk 1]
```diff
@@ -137,12 +137,16 @@ public XmlKeyManager(IOptions<KeyManagementOptions> keyManagementOptions, IActiv
 
     internal IXmlRepository KeyRepository { get; }
 
+    // Internal for testing
+    // Can't use TimeProvider since it's not available in framework
```
Comment (BrennanConroy) [inline src/DataProtection/DataProtection/src/KeyManagement/XmlKeyManager.cs:141]: https://www.nuget.org/packages/Microsoft.Bcl.TimeProvider

File: src/DataProtection/samples/KeyManagementSimulator/Program.cs [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+    object IActivator.CreateInstance(Type type, string _friendlyName) => type switch
+    {
+        Type t when t == typeof(IXmlDecryptor) => decryptor,
+        Type t when t == typeof(IAuthenticatedEncryptorDescriptorDeserializer) => descriptorDeserializer,
+        _ => throw new InvalidOperationException(),
+    };
+}
+
+/// <summary>
+/// A mock authenticated encryptor that doesn't only applies the identity function (i.e. does nothing).
```
Comment (BrennanConroy) [inline src/DataProtection/samples/KeyManagementSimulator/Program.cs:278]:
```suggestion
/// A mock authenticated encryptor that only applies the identity function (i.e. does nothing).
```

File: src/DataProtection/samples/KeyManagementSimulator/Program.cs [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+    /// <summary>
+    /// Trigger a key ring refresh and return enough information to update metrics
+    /// and advance the simulation.
+    /// </summary>
+    public KeyRingDescriptor RefreshKeyRingAndGetDescriptor(DateTimeOffset now)
+    {
+        // Update this before calling GetCacheableKeyRing so that it's available to XmlKeyManager
+        _now = now;
+
+        //Console.WriteLine($"[{_instanceNumber}] RefreshKeyRingAndGetDescriptor({now})");
```
Comment (BrennanConroy) [inline src/DataProtection/samples/KeyManagementSimulator/Program.cs:392]: Keeping these? Maybe have a debug flag you can pass in which turns on logging?

**Review by BrennanConroy — APPROVED** [review]

File: src/DataProtection/samples/KeyManagementSimulator/Program.cs (hunk 3 above)
Comment (amcasey) [reply src/DataProtection/samples/KeyManagementSimulator/Program.cs:392]: My thinking was that I was providing a sample of what one might want to log and that building in a verbosity system would be overkill.  I can get rid of them if you find them unhelpful.  Thoughts?

Comment (amcasey) [reply src/DataProtection/samples/KeyManagementSimulator/Program.cs]: Not worth revisiting - I've deleted those lines.

**Pull request merged**